File: core/modules/inventory/inventory.service.ts
/**
 * Inventory Service
 * Contains business logic for inventory operations
 */
import { IInventoryRepository } from './inventory.repository.interface.js';
import { VALID_INVENTORY_REASONS, INVENTORY_REASONS } from '../../../types/index.js';

let inventoryRepository: IInventoryRepository;
let productFlowService: any;
let auditService: any;
let transaction: any;

export function initInventoryService(
    repo: IInventoryRepository,
    pfService: any,
    audit: any,
    txFn: any
) {
    inventoryRepository = repo;
    productFlowService = pfService;
    auditService = audit;
    transaction = txFn;
}

/**
 * Get stock for a variant
 * @param {number} variantId - Variant ID
 * @returns {Object} Stock info
 */
export function getVariantStock(variantId: number) {
    const stock = inventoryRepository.getStockByVariantId(variantId);
    return { variantId, stock };
}

/**
 * Get inventory lots for a variant
 * @param {number} variantId - Variant ID
 * @returns {Array} Lots list
 */
export function getVariantLots(variantId: number) {
    return inventoryRepository.findLotsByVariantId(variantId);
}

/**
 * Get inventory adjustments for a variant
 * @param {number} variantId - Variant ID
 * @param {Object} options - Query options
 * @returns {Array} Adjustments list
 */
export function getVariantAdjustments(variantId: number, options: { limit?: number; offset?: number } = {}) {
    return inventoryRepository.findAdjustmentsByVariantId(variantId, options);
}

/**
 * Get all stock adjustments (paginated, searchable, filterable)
 */
export function getAllAdjustments(options: {
    limit?: number;
    offset?: number;
    search?: string;
    reason?: string;
    variantId?: number | null;
    sortBy?: string;
    sortOrder?: string;
} = {}) {
    return inventoryRepository.findAllAdjustments(options);
}

/**
 * Adjust inventory for a variant (creates adjustment record + product flow log)
 * @param {Object} params - Adjustment parameters
 * @returns {Object} Adjustment result
 */
export function adjustInventory({
    variantId,
    lotId,
    quantityChange,
    reason,
    referenceType,
    referenceId,
    userId
}: {
    variantId: number;
    lotId?: number | null;
    quantityChange: number;
    reason: string;
    referenceType?: string | null;
    referenceId?: number | null;
    userId: number;
}) {
    // Validate reason before transaction
    const validReasons = VALID_INVENTORY_REASONS as unknown as string[];
    if (!validReasons.includes(reason)) {
        throw new Error(`Invalid adjustment reason: ${reason}. Must be one of: ${validReasons.join(', ')}`);
    }

    // Validate stock availability BEFORE transaction for negative adjustments
    if (quantityChange < 0) {
        const currentStock = inventoryRepository.getStockByVariantId(variantId);
        if (currentStock + quantityChange < 0) {
            const error = new Error(`Insufficient stock. Available: ${currentStock}, Requested adjustment: ${quantityChange}`);
            (error as any).code = 'INSUFFICIENT_STOCK';
            throw error;
        }
    }

    const tx = transaction(() => {
        // Handle FIFO deduction if no lotId is provided for negative adjustment
        if (quantityChange < 0 && !lotId) {
            deductStockInternal({
                variantId,
                quantity: Math.abs(quantityChange),
                userId,
                reason,
                referenceType,
                referenceId
            });

            const newStock = inventoryRepository.getStockByVariantId(variantId);
            return {
                id: 0, // Multiple adjustments created
                variantId,
                quantityChange,
                reason,
                newStock
            };
        }

        // Create inventory adjustment
        const adjustmentResult = inventoryRepository.insertInventoryAdjustment({
            variant_id: variantId,
            lot_id: lotId,
            quantity_change: quantityChange,
            reason,
            reference_type: referenceType,
            reference_id: referenceId,
            adjusted_by: userId
        });

        const adjustmentId = Number(adjustmentResult.lastInsertRowid);
        const newStock = inventoryRepository.getStockByVariantId(variantId);

        // Queue cross-module operations for post-transaction execution
        const eventType = reason === INVENTORY_REASONS.SALE ? 'sale'
            : reason === INVENTORY_REASONS.RETURN ? 'return'
                : reason === INVENTORY_REASONS.CONFIRM_RECEIVE ? 'adjustment'
                    : 'adjustment';

        productFlowService.logProductFlow({
            variantId,
            eventType,
            quantity: quantityChange,
            referenceType: referenceType || 'adjustment',
            referenceId: referenceId || adjustmentId
        });

        auditService.logCreate(
            userId,
            'inventory_adjustments',
            adjustmentId,
            {
                variant_id: variantId,
                quantity_change: quantityChange,
                reason,
                new_stock: newStock
            }
        );

        return {
            id: adjustmentId,
            variantId,
            quantityChange,
            reason,
            newStock
        };
    });

    return tx.immediate();
}

/**
 * Get all variants with low stock
 * @returns {Array} Low stock variants
 */
export function getLowStockAlerts() {
    return inventoryRepository.findLowStockVariants();
}

/**
 * Get expiring lots
 * @param {number} days - Days to look ahead
 * @returns {Array} Expiring lots
 */
export function getExpiringLots(days: number = 30) {
    return inventoryRepository.findExpiringLots(days);
}

/**
 * Receive inventory from purchase order
 * Creates lots and adjustments for each item
 * @param {Object} params - Receipt parameters
 * @returns {Object} Receipt result
 */
export function receiveInventory({
    variantId,
    quantity,
    unitCost,
    batchNumber,
    manufacturedDate,
    expiryDate,
    purchaseOrderItemId,
    userId
}: {
    variantId: number;
    quantity: number;
    unitCost: number;
    batchNumber?: string | null;
    manufacturedDate?: string | null;
    expiryDate?: string | null;
    purchaseOrderItemId?: number | null;
    userId: number;
}) {
    const tx = transaction(() => {
        // Create inventory lot
        const lotResult = inventoryRepository.insertInventoryLot({
            variant_id: variantId,
            batch_number: batchNumber,
            manufactured_date: manufacturedDate,
            expiry_date: expiryDate,
            quantity,
            unit_cost: unitCost,
            purchase_order_item_id: purchaseOrderItemId
        });

        const lotId = Number(lotResult.lastInsertRowid);

        // Create adjustment for the receipt
        inventoryRepository.insertInventoryAdjustment({
            variant_id: variantId,
            lot_id: lotId,
            quantity_change: quantity,
            reason: INVENTORY_REASONS.CONFIRM_RECEIVE,
            reference_type: 'purchase_order_item',
            reference_id: purchaseOrderItemId,
            adjusted_by: userId
        });

        const newStock = inventoryRepository.getStockByVariantId(variantId);

        // Queue cross-module operations
        productFlowService.logProductFlow({
            variantId,
            eventType: 'purchase',
            quantity,
            referenceType: 'purchase_order_item',
            referenceId: purchaseOrderItemId
        });

        auditService.logCreate(
            userId,
            'inventory_lots',
            lotId,
            {
                variant_id: variantId,
                quantity,
                unit_cost: unitCost,
                batch_number: batchNumber,
                new_stock: newStock
            }
        );

        return {
            lotId,
            variantId,
            quantity,
            newStock
        };
    });

    return tx.immediate();
}

/**
 * Internal FIFO deduction logic without transaction wrapper
 */
function deductStockInternal({
    variantId,
    quantity,
    userId,
    reason,
    referenceType,
    referenceId
}: {
    variantId: number;
    quantity: number;
    userId: number;
    reason: string;
    referenceType?: string | null;
    referenceId?: number | null;
}) {
    let remainingToDeduct = quantity;
    const availableLots = inventoryRepository.findAvailableLots(variantId);

    for (const lot of availableLots) {
        if (remainingToDeduct <= 0) break;

        const deductionFromThisLot = Math.min(remainingToDeduct, lot.remaining_quantity);

        inventoryRepository.insertInventoryAdjustment({
            variant_id: variantId,
            lot_id: lot.id,
            quantity_change: -deductionFromThisLot,
            reason,
            reference_type: referenceType,
            reference_id: referenceId,
            adjusted_by: userId
        });

        remainingToDeduct -= deductionFromThisLot;
    }

    // If still remainingToDeduct > 0 (headless stock or negative stock allowed)
    if (remainingToDeduct > 0) {
        inventoryRepository.insertInventoryAdjustment({
            variant_id: variantId,
            lot_id: null,
            quantity_change: -remainingToDeduct,
            reason,
            reference_type: referenceType,
            reference_id: referenceId,
            adjusted_by: userId
        });
    }

    // Queue product flow logging
    const eventType = reason === INVENTORY_REASONS.SALE ? 'sale' : 'adjustment';
    productFlowService.logProductFlow({
        variantId,
        eventType,
        quantity: -quantity,
        referenceType: referenceType || 'adjustment',
        referenceId
    });

    return { success: true, deducted: quantity };
}

/**
 * Automatically deduct inventory stock from available lots (FIFO)
 * @param {Object} params - Deduction parameters
 * @returns {Object} Result
 */
export function deductStock(params: {
    variantId: number;
    quantity: number;
    userId: number;
    reason: string;
    referenceType?: string | null;
    referenceId?: number | null;
}) {
    if (params.quantity <= 0) return { success: true, deducted: 0 };
    const tx = transaction(() => deductStockInternal(params));
    return tx.immediate();
}

/**
 * Restore stock to lots by reversing previous adjustments
 */
export function restoreStock({
    variantId,
    referenceType,
    referenceId,
    quantity,
    userId,
    reason,
    newReferenceType,
    newReferenceId
}: {
    variantId: number;
    referenceType: string;
    referenceId: number;
    quantity: number;
    userId: number;
    reason: string;
    newReferenceType?: string | null;
    newReferenceId?: number | null;
}) {
    if (quantity <= 0) return { success: true, restored: 0 };

    const tx = transaction(() => {
        const originalAdjustments = inventoryRepository.findAdjustmentsByReference(referenceType, referenceId);
        let remainingToRestore = quantity;

        // Sort by adjusted_at DESC to reverse LIFO (reverse of FIFO deduction)
        originalAdjustments.sort((a, b) => new Date(b.adjusted_at).getTime() - new Date(a.adjusted_at).getTime());

        for (const adj of originalAdjustments) {
            if (remainingToRestore <= 0) break;

            const originalDeduction = Math.abs(adj.quantity_change);
            const restoreToThisLot = Math.min(remainingToRestore, originalDeduction);

            inventoryRepository.insertInventoryAdjustment({
                variant_id: variantId,
                lot_id: adj.lot_id,
                quantity_change: restoreToThisLot,
                reason,
                reference_type: newReferenceType,
                reference_id: newReferenceId,
                adjusted_by: userId
            });

            remainingToRestore -= restoreToThisLot;
        }

        // If still remaining (headless stock or something went wrong)
        if (remainingToRestore > 0) {
            inventoryRepository.insertInventoryAdjustment({
                variant_id: variantId,
                lot_id: null,
                quantity_change: remainingToRestore,
                reason,
                reference_type: newReferenceType,
                reference_id: newReferenceId,
                adjusted_by: userId
            });
        }

        // Queue product flow logging
        const eventType = reason === INVENTORY_REASONS.RETURN ? 'return' : 'adjustment';
        productFlowService.logProductFlow({
            variantId,
            eventType,
            quantity: quantity,
            referenceType: newReferenceType || 'adjustment',
            referenceId: newReferenceId
        });

        return { success: true, restored: quantity };
    });

    return tx.immediate();
}

/**
 * Get aggregate inventory stats
 * @returns {Object} Inventory stats
 */
export function getInventoryStats() {
    return inventoryRepository.getInventoryStats();
}

File: src/main/java/com/possum/application/inventory/InventoryService.java
package com.possum.application.inventory;

import com.possum.domain.services.StockManager;

import com.possum.domain.enums.FlowEventType;
import com.possum.domain.enums.InventoryReason;
import com.possum.domain.exceptions.InsufficientStockException;
import com.possum.domain.model.InventoryAdjustment;
import com.possum.domain.model.InventoryLot;
import com.possum.domain.model.Variant;
import com.possum.infrastructure.filesystem.SettingsStore;
import com.possum.infrastructure.serialization.JsonService;
import com.possum.persistence.db.TransactionManager;
import com.possum.domain.repositories.AuditRepository;
import com.possum.domain.repositories.InventoryRepository;
import com.possum.shared.dto.AvailableLot;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import com.possum.shared.util.TimeUtil;
import java.util.List;
import java.util.Map;

public class InventoryService {
    private final InventoryRepository inventoryRepository;
    private final ProductFlowService productFlowService;
    private final AuditRepository auditRepository;
    private final TransactionManager transactionManager;
    private final JsonService jsonService;
    private final SettingsStore settingsStore;
    private final StockManager stockManager;

    public InventoryService(InventoryRepository inventoryRepository,
                            ProductFlowService productFlowService,
                            AuditRepository auditRepository,
                            TransactionManager transactionManager,
                            JsonService jsonService,
                            SettingsStore settingsStore,
                            StockManager stockManager) {
        this.inventoryRepository = inventoryRepository;
        this.productFlowService = productFlowService;
        this.auditRepository = auditRepository;
        this.transactionManager = transactionManager;
        this.jsonService = jsonService;
        this.settingsStore = settingsStore;
        this.stockManager = stockManager;
    }

    public int getVariantStock(long variantId) {
        return inventoryRepository.getStockByVariantId(variantId);
    }

    public List<InventoryLot> getVariantLots(long variantId) {
        return inventoryRepository.findLotsByVariantId(variantId);
    }

    public List<InventoryAdjustment> getVariantAdjustments(long variantId, int limit, int offset) {
        return inventoryRepository.findAdjustmentsByVariantId(variantId, limit, offset);
    }

    public List<com.possum.shared.dto.StockHistoryDto> getStockHistory(String search, List<String> reasons, String fromDate, String toDate, List<Long> userIds, int limit, int offset) {
        return inventoryRepository.findStockHistory(search, reasons, fromDate, toDate, userIds, limit, offset);
    }

    public List<Variant> getLowStockAlerts() {
        return inventoryRepository.findLowStockVariants();
    }

    public List<InventoryLot> getExpiringLots(int days) {
        return inventoryRepository.findExpiringLots(days);
    }

    public Map<String, Object> getInventoryStats() {
        return inventoryRepository.getInventoryStats();
    }

    public ReceiveInventoryResult receiveInventory(long variantId, int quantity, BigDecimal unitCost,
                                                   String batchNumber, LocalDateTime manufacturedDate,
                                                   LocalDateTime expiryDate, Long purchaseOrderItemId, long userId) {
        if (quantity <= 0)
            throw new com.possum.domain.exceptions.ValidationException("Quantity must be positive");
        if (unitCost == null || unitCost.compareTo(BigDecimal.ZERO) < 0)
            throw new com.possum.domain.exceptions.ValidationException("Unit cost must be zero or greater");

        return transactionManager.runInTransaction(() -> {
            InventoryLot lot = new InventoryLot(null, variantId, batchNumber, manufacturedDate, expiryDate,
                    quantity, unitCost, purchaseOrderItemId, TimeUtil.nowUTC());
            long lotId = inventoryRepository.insertInventoryLot(lot);

            InventoryAdjustment adjustment = new InventoryAdjustment(null, variantId, lotId, quantity,
                    InventoryReason.CONFIRM_RECEIVE.getValue(), "purchase_order_item", purchaseOrderItemId,
                    userId, null, TimeUtil.nowUTC());
            inventoryRepository.insertInventoryAdjustment(adjustment);

            int newStock = inventoryRepository.getStockByVariantId(variantId);

            productFlowService.logProductFlow(variantId, FlowEventType.PURCHASE, quantity,
                    "purchase_order_item", purchaseOrderItemId);

            Map<String, Object> auditData = Map.of(
                    "variant_id", variantId,
                    "quantity", quantity,
                    "unit_cost", unitCost,
                    "batch_number", batchNumber != null ? batchNumber : "",
                    "new_stock", newStock
            );
            auditRepository.log("inventory_lots", lotId, "CREATE", jsonService.toJson(auditData), userId);

            return new ReceiveInventoryResult(lotId, variantId, quantity, newStock);
        });
    }

    public DeductStockResult deductStock(long variantId, int quantity, long userId, InventoryReason reason,
                                         String referenceType, Long referenceId) {
        if (quantity <= 0) {
            return new DeductStockResult(true, 0);
        }

        int currentStock = inventoryRepository.getStockByVariantId(variantId);
        if (isInventoryRestrictionsEnabled() && currentStock < quantity) {
            throw new InsufficientStockException(currentStock, quantity);
        }

        return transactionManager.runInTransaction(() -> {
            deductStockInternal(variantId, quantity, userId, reason, referenceType, referenceId);
            return new DeductStockResult(true, quantity);
        });
    }

    public RestoreStockResult restoreStock(long variantId, String referenceType, long referenceId,
                                           int quantity, long userId, InventoryReason reason,
                                           String newReferenceType, Long newReferenceId) {
        if (quantity <= 0) {
            return new RestoreStockResult(true, 0);
        }

        return transactionManager.runInTransaction(() -> {
            List<InventoryAdjustment> originalAdjustments = inventoryRepository
                    .findAdjustmentsByReference(referenceType, referenceId);

            List<InventoryAdjustment> plan = stockManager.planRestoration(
                    variantId, quantity, originalAdjustments, reason, newReferenceType, newReferenceId, userId
            );

            for (InventoryAdjustment adj : plan) {
                inventoryRepository.insertInventoryAdjustment(adj);
            }

            FlowEventType eventType = reason == InventoryReason.RETURN ? FlowEventType.RETURN : FlowEventType.ADJUSTMENT;
            productFlowService.logProductFlow(variantId, eventType, quantity,
                    newReferenceType != null ? newReferenceType : "adjustment", newReferenceId);

            return new RestoreStockResult(true, quantity);
        });
    }

    public AdjustInventoryResult adjustInventory(long variantId, Long lotId, int quantityChange,
                                                 InventoryReason reason, String referenceType,
                                                 Long referenceId, long userId) {
        if (quantityChange < 0 && isInventoryRestrictionsEnabled()) {
            int currentStock = inventoryRepository.getStockByVariantId(variantId);
            if (currentStock + quantityChange < 0) {
                throw new InsufficientStockException(currentStock, Math.abs(quantityChange));
            }
        }

        return transactionManager.runInTransaction(() -> {
            if (quantityChange < 0 && lotId == null) {
                deductStockInternal(variantId, Math.abs(quantityChange), userId, reason, referenceType, referenceId);
                int newStock = inventoryRepository.getStockByVariantId(variantId);
                return new AdjustInventoryResult(0, variantId, quantityChange, reason, newStock);
            }

            InventoryAdjustment adjustment = new InventoryAdjustment(null, variantId, lotId, quantityChange,
                    reason.getValue(), referenceType, referenceId, userId, null, TimeUtil.nowUTC());
            long adjustmentId = inventoryRepository.insertInventoryAdjustment(adjustment);

            int newStock = inventoryRepository.getStockByVariantId(variantId);

            FlowEventType eventType = switch (reason) {
                case SALE -> FlowEventType.SALE;
                case RETURN -> FlowEventType.RETURN;
                default -> FlowEventType.ADJUSTMENT;
            };

            productFlowService.logProductFlow(variantId, eventType, quantityChange,
                    referenceType != null ? referenceType : "adjustment", referenceId != null ? referenceId : adjustmentId);

            Map<String, Object> auditData = Map.of(
                    "variant_id", variantId,
                    "quantity_change", quantityChange,
                    "reason", reason.getValue(),
                    "new_stock", newStock
            );
            auditRepository.log("inventory_adjustments", adjustmentId, "CREATE", jsonService.toJson(auditData), userId);

            return new AdjustInventoryResult(adjustmentId, variantId, quantityChange, reason, newStock);
        });
    }

    private void deductStockInternal(long variantId, int quantity, long userId, InventoryReason reason,
                                     String referenceType, Long referenceId) {
        List<AvailableLot> availableLots = inventoryRepository.findAvailableLots(variantId);
        List<InventoryAdjustment> plan = stockManager.planDeduction(
                variantId, quantity, availableLots, reason, referenceType, referenceId, userId
        );

        for (InventoryAdjustment adj : plan) {
            inventoryRepository.insertInventoryAdjustment(adj);
        }

        FlowEventType eventType = reason == InventoryReason.SALE ? FlowEventType.SALE : FlowEventType.ADJUSTMENT;
        productFlowService.logProductFlow(variantId, eventType, -quantity,
                referenceType != null ? referenceType : "adjustment", referenceId);
    }

    public record ReceiveInventoryResult(long lotId, long variantId, int quantity, int newStock) {}
    public record DeductStockResult(boolean success, int deducted) {}
    public record RestoreStockResult(boolean success, int restored) {}
    public record AdjustInventoryResult(long id, long variantId, int quantityChange, InventoryReason reason, int newStock) {}

    private boolean isInventoryRestrictionsEnabled() {
        try {
            return settingsStore.loadGeneralSettings().isInventoryAlertsAndRestrictionsEnabled();
        } catch (Exception ex) {
            return true;
        }
    }
}
